assembly/src/compile.rs

```diff
@@ -0,0 +1,332 @@
+use alloc::{
+    borrow::Cow,
+    boxed::Box,
+    string::{String, ToString},
+    sync::Arc,
+    vec::Vec,
+};
+
+use crate::{
+    ast::{Module, ModuleKind},
+    diagnostics::{
+        Diagnostic, IntoDiagnostic, NamedSource, Report, SourceCode, SourceFile, WrapErr,
+    },
+    library::{LibraryNamespace, LibraryPath},
+};
+
+// COMPILE OPTIONS
+// ================================================================================================
+
+/// The set of options which can be used to control the behavior of the [Compile] trait.
+#[derive(Debug, Clone)]
+pub struct Options {
+    /// The kind of [Module] to compile.
+    ///
+    /// The default kind is executable.
+    pub kind: ModuleKind,
+    /// The namespace to apply to the compiled [Module].
+    ///
+    /// If unset, the namespace will be derived from the [ModuleKind].
+    pub namespace: Option<LibraryNamespace>,
+    /// The name to give the compiled [Module]
+    ///
+    /// This option overrides `namespace`.
+    ///
+    /// If unset, and there is no name associated with the item being compiled (e.g. a file path)
+    /// then the path will consist of just a namespace; using the value of `namespace` if provided,
+    /// or deriving one from `kind`.
+    pub path: Option<LibraryPath>,
```

Comment:
Why do we need to have both `namespace` and `path`? Having `LibraryPath` implies `namespace` - no? So, we could probably just keep `path` and if we wanted to get access to the namespace we could just get it from the path.

Also, would it make sense to make these non-optional and populate with default values in the constructor? This way, the code that handles optionality can be encapsulated here rather then being in a few places.

For example, it could be something like:
```Rust
pub struct Options {
    kind: ModuleKind,
    path: LibraryPath,
}

impl Options {

    pub fn from_kind(kind: ModuleKind) -> Self {
        let path = LibraryPath::from(match options.kind {
            ModuleKind::Library => LibraryNamespace::Anon,
            ModuleKind::Executable => LibraryNamespace::Exec,
            ModuleKind::Kernel => LibraryNamespace::Kernel,
        });
        Self { kind, path }
    }

    pub fn kind(&self) -> ModuleKind { ... }
    pub fn path(&self) -> &LibraryPath { ... }
    pub fn namespace(&self) -> &LibraryNamespace { ... }
}
```

Reply:
The `namespace` option is no longer needed due to some refactorings I did after this was written, but didn't remove it for some reason (I guess I must've thought it could come in handy in tests or something).

In any case, `path` is the only one we actually need/use.

> Also, would it make sense to make these non-optional and populate with default values in the constructor?

Not in this case, since an implementation of `Compile` could be an already-compiled `Module`, in which case we do not want to override the path unless that is truly what is wanted.

Similarly, there are cases where the path might be derived from the `SourceFile`, and we don't want to override that either, unless it is specifically the case that we want to ignore it and provide a different one.